fix: 2881 akconcatenate fails trying to concatenate too many nested arrays

tcawlfield opened the pull request:
When concatenating arrays at a depth > 0 (not outermost), current code has a limitation
of 128 arrays to be combined.

This PR relaxes that requirement.

tcawlfield committed:
fix: first stab at 2881. WIP

This breaks a bunch of old tests, but passes the
original test in 2881. Needs scrutiny etc.

tcawlfield committed:
New approach: add partial support for 64-bit tags

Expanding the template instances for:
* awkward_UnionArray_nestedfill_tags_index
* awkward_UnionArray_simplify_one
* awkward_UnionArray_simplify
These allow 64-bit tags in certain cases, for ak_concatenate.

jpivarski commented on src/awkward/contents/unionarray.py:
```diff
@@ -394,6 +394,7 @@ def simplified(
                     contents.append(self_cont)
 
         if len(contents) > 2**7:
+            # TODO: Catch this condition in the loops above.
```
Don't forget this TODO!

jpivarski commented on src/awkward/operations/ak_concatenate.py:
```diff
@@ -152,7 +152,7 @@ def _impl(arrays, axis, mergebool, highlevel, behavior, attrs):
         for x in content_or_others[1:]:
             batch = batches[-1]
-            if ak._do.mergeable(batch[-1], x, mergebool=mergebool):
+            if ak._do.mergeable(batch[0], x, mergebool=mergebool):
```
```suggestion
            if ak._do.mergeable(batch[-1], x, mergebool=mergebool):
```

Although `mergeable` is an equivalence relation and we can compare a new item with any of the already established items, I think it's easier to think about as always checking the last one. (If you were to draw a diagram of it, the lines wouldn't cross.)

jpivarski commented on src/awkward/operations/ak_concatenate.py:
```diff
@@ -302,7 +308,7 @@ def action(inputs, depth, backend, **kwargs):
 
                 for x in nextinputs:
-                    o, f = x._offsets_and_flattened(1, 1)
+                    o, f = x._offsets_and_flattened(1, 1)  # axis=1, depth=1
```
```suggestion
                    o, f = x._offsets_and_flattened(axis=1, depth=1)
```

Adding this comment is helpful, but calling with keyword arguments can't get out of date. This is a bit of code that was originally translated from C++, so that's why it was like this.

tcawlfield committed:
Ordering specializations in kernel-specification.yml

jpivarski reviewed (approved):
This is looking good!

One thing that's missing is for `UnionArray.simplified` to document that it accepts 64-bit tags, but if the final result would be to make a `UnionArray` with more than more than 127 `contents`, it still raises an error. (And also, that check needs to be implemented.)

tcawlfield committed:
UnionArray.simplify length tests moved.

Also a few style modifications.
The reason for moving those length checks
is so that I do not wish even temporarily hold on
to content tags with invalid data.

tcawlfield committed:
Reverting change to batch index in ak_concatenate

Co-authored-by: Jim Pivarski <jpivarski@users.noreply.github.com>

ianna reviewed (approved):
@tcawlfield - looks great to me! All CUDA tests pass as expected! Thanks!
